specs/0118-notification-center.md

```diff
@@ -0,0 +1,119 @@
+- Feature Name: Notification Center
+- Created: 2024-03-12
+- Associated Issue: 
+
+# Summary
+
+Introduce a Notification Center feature to make Pontoon the best tool to send communications to localizers directly, allowing sending of targeted in-app notifications and emails within Pontoon itself.
+
+# Motivation
+
+Currently, we can only send in-app notifications with limited targeting capabilities (selected locales within a single product). For any other communications targeting a subset of users we need to rely on engineers to extract email addresses and send emails to the targeted recipients. This feature would allow administrators of a Pontoon instance managing localization projects to send notifications or emails to targeted sets of recipients with more refined criteria such as locales, projects, roles, recency of activity, and level of activity.
+
+# Feature explanation
+
+## Notification Center
+
+This feature has the ability to compose a message with a subject and body, send a message as either an in-app notification or email, and target a set of recipients based on commonly used criteria.
+
+### Required functionality:
+
+- Title: `Notification Center`
+- Header: `Message type`
+  - Selects the type of message to be sent
+  - Checkboxes: `Email`, `Notification` (multiselect, default: both boxes empty)
+  - Validation: Error if no checkbox selected - `You must select at least one message type.`
+- Header: `Message editor`
+  - Creates a subject and body for emails. Subject line will be required for in-app messages; in app-messages have the subject line combined with the body message as the first line of the message separated by a newline.
+  - Field: `Subject` (text input)
+  - Field: `Body` (text input with formatting: markdown or html)
+  - Button: `Preview` - shows a rendered version of the text in `Body` for checking formatting, similar to how the preview for GitHub comments work
+  - Validation: 
+    - Error if `Subject` field empty: `Your message must include a subject.`
+    - Error if `Body` field empty: `Your message must include a body.`
+- Header: `Recipients`
+  - Select who will receive a message based on a combination of the below filters.
+  - Checkbox: `Filter by User Role` (default empty) - enables User Role as a filter.
```

Comment:
That's how I had originally planned it. I'm fine either way, but I suggested this when creating the spec as it would mean constantly having to click all the user roles. remembering to move all the locales / projects to chosen if we didn't want to filter by these.
With a checkbox we the could enable them only when applicable.

Reply:
+1 on using checkboxes on the "rule". If we don't want to filter by user role, we don't select the specific rule, and target all of them. That's a lot faster and, at least for me, clearer.

P.S. empty ⇾ deselected.

Reply:
I have a few concerns about checkboxes in the current version of the spec:
* You can select "Filter by Role" and then don't select any Role. Not great UX, which requires validation.
* Some headers have checkboxes, some don't. It's not clear why the difference.
* The current version of the spec suggests lots of checkboxes (10).
* Having all checkboxes selected yields the same results as as not having any checkbox selected.

_This is the default state of the User Role filters according to the spec:_
<img width="204" alt="Screenshot 2024-03-29 at 18 21 53" src="https://github.com/mozilla/pontoon/assets/626716/1d052ba9-8d0b-4d8d-b35b-74b4b5ca74b5">

_This is the default state of the User Role filters without the checkbox in the header:_
<img width="163" alt="Screenshot 2024-03-29 at 18 19 37" src="https://github.com/mozilla/pontoon/assets/626716/e956865e-b8f4-45a8-92dc-e2ba0e3ce89b">

> +1 on using checkboxes on the "rule". If we don't want to filter by user role, we don't select the specific rule, and target all of them. That's a lot faster and, at least for me, clearer.

To target all roles, you don't do anything in either case, so both options are equally fast.

Are empty checkboxes in front of all user roles really making it clearer that you're targeting all roles?

Reply:
It's clearer for me, because I'm effectively choosing not to "filter by user role" if the checkbox is unchecked (and the checkboxes for each user role would be disabled in that case). Same for other rules: if it's not selected, the associated controls would be disabled. 

With that said, as long as the functionality is there, I can live with either solutions, also considering the users are internal and just need to learn how to use the feature. I also see how the second requires less validation.

Looking at your screenshots, I think we need something to clarify that the selected roles are included when selected (especially in the second). Filter is ambiguous (include or exclude?).

Reply:
I've eliminated the checkboxes and changed `Filter by user role` to `Included User Roles`.